Resolves #49578 ('Delete holes' bug when 'area' argument is used)

TomasJohansson opened the pull request:
## Description

Fixed the problem in issue [49578](https://github.com/qgis/QGIS/issues/49578)

But the third commit is just a performance improvement for a method I noticed when fixing the problem,
and I thought it would be considered as overkill to create a separate pull request for that commit (6e8de5cd164b0cdc2fb67a470fad8fe06fd3e725)



<!--
  BEFORE HITTING SUBMIT -- Please BUILD AND TEST your changes thoroughly. This is YOUR responsibility! Do NOT rely on the QGIS code maintainers to do this for you!!

  IMPORTANT NOTES FOR FIRST TIME CONTRIBUTORS
  ===========================================

  Congratulations, you are about to make a pull request to QGIS! To make this as easy and pleasurable for everyone, please take the time to read these lines before opening the pull request.

  Include a few sentences describing the overall goals for this pull request (PR). If applicable also add screenshots or - even better - screencasts.
  Include both: *what* you changed and *why* you changed it.

  If this is a pull request that adds new functionality which needs documentation, give an especially detailed explanation.
  In this case, start with a short abstract and then write some text that can be copied 1:1 to the documentation in the best case.

  Also mention if you think this PR needs to be backported. And list relevant or fixed issues.

------------------------

  Reviewing is a process done by project maintainers, mostly on a volunteer basis. We try to keep the overhead as small as possible and appreciate if you help us to do so by checking the following list.
  Feel free to ask in a comment if you have troubles with any of them.

  - Commit messages are descriptive and explain the rationale for changes.

  - Commits which fix bugs include `Fixes #11111` at the bottom of the commit message. If this is your first pull request and you forgot to do this, write the same statement into this text field with the pull request description.

  - New unit tests have been added for relevant changes

  - You have run the `scripts/prepare_commit.sh` script (https://github.com/qgis/QGIS/blob/master/.github/CONTRIBUTING.md#contributing-to-qgis) before each commit.
    If you didn't do this, you can also run `./scripts/astyle_all.sh` from your source folder.

  - You have read the QGIS Coding Standards (https://docs.qgis.org/testing/en/docs/developers_guide/codingstandards.html) and this PR complies with them
-->

Fixes #49578

TomasJohansson committed:
Bugfixed 'Delete holes' (removeInteriorRings) when the area argument is used. Github issue #49578

TomasJohansson committed:
Refactored the test code for github issue 49578.
Removed some duplication by using two lambda functions,
and added some variables (constants) instead of repeating so called 'magic numbers' for the areas.
(also removed a comment added in the previous commit)

TomasJohansson committed:
Refactoring. Removed unnecessary repeated multiplications within a loop,
potentially iterating thousands of points (i.e. that many unnecessary multiplications).
I did read the message that the class is critical and changes must be accompanied with unit tests,
but I do not now add any new tests since there are already tests that detects an introduced error.
Before this commit, I tested with another value than 0.5 after the loop, and
then indeed some tests failed ('TestQgsCurvePolygon::removeInteriorRings_github_issue_49578()' and 'TestQgsLineString::sumUpArea()')
but when now using 0.5 those tests are succeeding.

TomasJohansson committed:
Changed the british english spelling of 'behaviour' to the american english 'behavior' because of the 'failing check' https://github.com/qgis/QGIS/runs/7685754342?check_suite_focus=true
for the pull request https://github.com/qgis/QGIS/pull/49583

nyalldawson commented on src/core/geometry/qgslinestring.cpp:
```diff
@@ -1936,12 +1936,16 @@ QgsPoint QgsLineString::centroid() const
 
 void QgsLineString::sumUpArea( double &sum ) const
 {
+  // Please note that the area calculated in this method will be negative if the line points are defined in clockwise order.
```
These comments would belong better as additions to the docstrings in the header file, so that they get included in the API docs and PyQGIS references.

TomasJohansson replied:
OK I have now added comments to the header file instead.

TomasJohansson committed:
Added comments to the header file instead of the cpp file, as suggested in the pull request feedback https://github.com/qgis/QGIS/pull/49583#discussion_r939593562
Also recreated a corresponding sip file as described here:
https://docs.qgis.org/3.22/en/docs/developers_guide/codingstandards.html#generating-the-sip-file

Regarding the URL I added with "\see" in the comment in the header file, I did it as in an example I found at the method "parseLinkTypes" (in the file qgsmetadatawidget.h).
I noted that the generated sip file does not include the URL but only "https", and that does not seem to be a new problem but is the same for the sip file "qgsmetadatawidget.sip.in" (and thus the same problem in the generated html https://qgis.org/pyqgis/3.24/gui/QgsMetadataWidget.html)

TomasJohansson committed:
Generated a new version of the file "qgslinestring.sip.in"
(since I noticed a sip_check failure https://github.com/qgis/QGIS/runs/7830097236?check_suite_focus=true)

TomasJohansson commented:
Today I noted that there is now a new result but with one failure.
I do not know how the process works, i.e. maybe someone is now waiting for me to commit something new to fix the problem 
and then after a new expected commit, someone is automatically notified about it and then decides to try executing the tests again?

If so, I now want to say that you should not wait for me to do that, since I am pretty sure that this test failure has nothing to do with this pull request of mine.
When I clicked "Details" of the failed tests it leads to this URL:
https://github.com/qgis/QGIS/runs/7834878518?check_suite_focus=true
```
The following tests FAILED:
	136 - test_core_networkaccessmanager (Timeout)
```

I have not commited anything related to network access.

nyalldawson commented:
@TomasJohansson the merge conflict needs fixing here -- it's preventing the ci tests from running

TomasJohansson committed:
Merge branch 'master' into delete-holes-removeInteriorRings

TomasJohansson commented:
@nyalldawson Okay, now I have merged

TomasJohansson commented:
I merged 15 days ago.
"Review required" and "5 workflows awaiting approval"
@nyalldawson

TomasJohansson committed:
Merge branch 'master' into delete-holes-removeInteriorRings

# Conflicts:
#	resources/data/contributors.json
#	src/core/geometry/qgslinestring.cpp

TomasJohansson committed:
Merge branch 'master' into delete-holes-removeInteriorRings

TomasJohansson committed:
Merge branch 'delete-holes-removeInteriorRings2' into delete-holes-removeInteriorRings

TomasJohansson commented:
Today I again had to merge.
There had been changes to qgslinestring.cpp which I now let override my performance improvement I mentioned when I created this pull request **almost two months ago**.
So, now there is only the following **one line of production code** (the other things in this PR are just test code and non-executable things such as comments) that needs to be merged into the master by someone with write access:
(in the file qgscurvepolygon.cpp)
`if ( area < minimumAllowedArea )`
==> 
`if ( std::fabs( area ) < minimumAllowedArea )`

nyalldawson commented:
@TomasJohansson the fix is good, sorry about the slow merge! 

I think it's worth reviving the optimisation (the unnecessary duplicate multiplication) , but can we do that in a follow up pr?

TomasJohansson commented:
> @TomasJohansson the fix is good, sorry about the slow merge!
> 
> I think it's worth reviving the optimisation (the unnecessary duplicate multiplication) , but can we do that in a follow up pr?

Thank you, okay I have now created a [new PR for the duplicate multiplications](https://github.com/qgis/QGIS/pull/50450)
